### PR TITLE
refactor code to follow standard layout and make form in switch page

### PR DESCRIPTION
**Pull request by itsibitzi**
Change the switch toggling to use a form instead of a bare basic auth'd POST request.

Most the rest of the diff is refactoring to: 
* Remove the per-switch action builder.
* Generalise the switch interface. We had generalised switch definitions but static endpoints which was weird.
* Adding CSRF tokens to POST endpoints. I have no idea how the email one worked before...
* Updated readme to reflect new changes

### CODE REVIEW

File: app/model/SwitchError.scala [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import play.api.mvc.Result
+import play.api.mvc.Results._
+
+sealed trait SwitchError {
+  def toResult(message: String): Result = this match {
+    case SwitchError.UsersTableError  => InternalServerError(message)
+    case SwitchError.UserNotFoundError  => Unauthorized(message)
+    case SwitchError.AccessBadPasswordError => Unauthorized(message)
+    case SwitchError.PutSwitchStatusError => InternalServerError(message)
+    case SwitchError.NonWhitelistedSwitchError(switch) => BadRequest(s"Non-whitelisted switch: $switch")
```
[the remark this answers is not in the document]
Comment (itsibitzi) [reply app/model/SwitchError.scala]: Black lists list everything you don't want, white lists list what you do want, non-white listed is everything that you haven't said you want.

At the moment our black list have would be every possible string in the universe except `emergency`. :P

[the remark this answers is not in the document]
Comment (akash1810) [reply app/model/SwitchError.scala]: minor: Would `UnrecognisedSwitchError` be more descriptive here? (unless I'm alone in treating black/white lists as binary).

[the remark this answers is not in the document]
Comment (sihil) [reply app/model/SwitchError.scala]: 👍 for "unrecognised switch".

File: app/controllers/SwitchesController.scala [hunk 2]
```diff
@@ -1,26 +1,28 @@
...
-    }.getOrElse(InternalServerError(views.html.switches.switchChange(errorMessage(On))))
-  }
-
-  def emergencyOff = EmergencySwitchChangeAccess { req =>
-    switches.setEmergencySwitch(Off).map { _ =>
-      Ok(views.html.switches.switchChange(success(Off)))
-    }.getOrElse(InternalServerError(views.html.switches.switchChange(errorMessage(Off))))
+  def setSwitch(switchName: String) = Action(parse.formUrlEncoded) { req =>
+    (for {
+        username <- req.body.get("username").flatMap(_.headOption)
```
[the remark this answers is not in the document]
Comment (sihil) [reply app/controllers/SwitchesController.scala]: `body` is a Map[String, Seq[String]] here. I quite like this pattern of extracting form fields (its easier to grok than Play's magic and doesn't seem to be any more verbose). I would however try to improve how error messaging is done - perhaps by making the for comp an Either[Result, X] monad that can shortcircuit to a `Left(BadRequest("username field missing"))`.

**Comment by sihil** [comment]
I've rebased this for you @itsibitzi. You will now find you can deploy to CODE and test it 🎈

**Pull request closed without merging**